File: week01/cash/cash.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int number_of_coins = 0;
float get_float_above(int number, string prompt);
int count_number_of_coins(int change_in_cents, int coin_size);

int main(void)
{
    float change_in_dollars = get_float_above(0, "Change owed:");
    int change_in_cents = round(change_in_dollars * 100);

    change_in_cents = count_number_of_coins(change_in_cents, 25);
    change_in_cents = count_number_of_coins(change_in_cents, 10);
    change_in_cents = count_number_of_coins(change_in_cents, 5);
    change_in_cents = count_number_of_coins(change_in_cents, 1);

    printf("%i\n", number_of_coins);
}

float get_float_above(int number, string prompt)
{
    float n;
    do
    {
        n = get_float("%s ", prompt);
    }
    while (n <= number);
    return n;
}

int count_number_of_coins(int change_in_cents, int coin_size)
{
    while ( coin_size <= change_in_cents )
    {
        change_in_cents = change_in_cents - coin_size;
        ++number_of_coins;
    }
    return change_in_cents;
}

File: week06/cash/cash.py
from cs50 import get_float


COIN_SIZES = [25, 10, 5, 1]


def main():
    change = get_change()
    change = convert_change_to_cents(change)
    number_of_coins = calc_total_number_of_coins(change)
    print(number_of_coins)


def get_change():
    change = get_float("Change owed: ")
    while change < 0:
        change = get_float("Change owed: ")
    
    return change


def convert_change_to_cents(change):
    return int(round(change * 100))


def calc_total_number_of_coins(change):
    number_of_coins = 0 
    for coin_size in COIN_SIZES:
        # Count new coins
        counted_coins = calc_number_of_coins_with_size(change, coin_size)
        # Add new coins to total number of coins
        number_of_coins += counted_coins
        # Calculate remaining change
        change -= counted_coins * coin_size
    
    return number_of_coins


def calc_number_of_coins_with_size(change, coin_size):
    number_of_coins = 0
    while coin_size <= change:
        number_of_coins += 1
        change -= coin_size
    
    return number_of_coins


main()
